include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h

```diff
@@ -0,0 +1,165 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2023 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#ifndef EXADG_ACOUSTIC_CONSERVATION_EQUATIONS_SPATIAL_DISCRETIZATION_OPERATORS_WEAK_BOUNDARY_CONDITIONS_H_
+#define EXADG_ACOUSTIC_CONSERVATION_EQUATIONS_SPATIAL_DISCRETIZATION_OPERATORS_WEAK_BOUNDARY_CONDITIONS_H_
+
+#include <exadg/acoustic_conservation_equations/user_interface/boundary_descriptor.h>
+#include <exadg/functions_and_boundary_conditions/evaluate_functions.h>
+#include <exadg/matrix_free/integrators.h>
+
+namespace ExaDG
+{
+namespace Acoustics
+{
+/**
+ * Class to access boundary values with a similar interface to FaceIntegrators.
```

Comment:
I agree only partly on the statement that the interface is similar to `FaceIntegrator`. I think we should be more explicit here in the sense that we need generic programming regarding the two functions `reinit()`/`get_value()`, not more than this.

I am not fully happy with the current design. Most of the code of this class is duplicated between `BoundaryFaceIntegratorP` and `BoundaryFaceIntegratorU`, which should be avoided. The only difference is the implementation of `get_value()`, which is problem/operator-specific and where we have free functions in ExaDG. The use of free functions is a modular design in the sense that it keeps dependencies to a minimum. ExaDG implementations for other operators are more general in the sense of an additional parameter `operator_type` (https://github.com/exadg/exadg/blob/ee073c7979b3c695aca00ade2ab9ecfebc74a410/include/exadg/incompressible_navier_stokes/spatial_discretization/operators/weak_boundary_conditions.h#L72) needed for implicit time stepping, which does not appear here for the explicit acoustic operator, which always evaluates the `full` operator.


There are two open questions in my opinion:
- Why not introducing a base class `BoundaryFaceIntegratorBase`?
- How to realize `get_value()` in a generic way to be able to deal with all types of ExaDG operators? I see the three options of (i) templating `BoundaryFaceIntegratorBase` with `BoundaryDescriptor`, (ii) using problem specific derived classes `BoundaryFaceIntegratorDerived` (where an concrete boundary descriptor is used but which does not allow the short-cut in the `reinit()` function used currently), or (iii) the use of lambda functions in order to hide `BoundaryDescriptor` (and other problem-specific parameters).

Reply:
As written here https://github.com/exadg/exadg/pull/587/files#r1383107143, the "interface" / "similarity" with `FEFaceEvaluation` seems to be `get_value()` only.

Reply:
A question would then indeed be whether it makes sense that dealii  provides a base/interface class `FEFaceEvaluationInterface` with `get_value()`/`get_gradient()` only. `FEFaceIntegrationInterface` as a derived class would add the `submit()` functionality ("Integration" = testing, weighting) to `FEFaceEvaluationInterface` according to this idea.

Reply:
I didn't introduce a base class, since the amount of duplicated code is small. If this requested base class will be used in other parts of `ExaDG` in the future, I agree that we should have one. I will restructure as suggested, just a few answers to your comments:

* I think the description is correct: The interface **to access the value** is similar to `FEEval`.
* Obviously additional parameters such as `OperatorType` etc. can be easily added, but I think this is not relevant for this PR and should be done when all the other operators are adjusted to this design.
* I am in favor of templating the boundary descriptor. It is natural to fetch the boundary ID in the `reinit()` function and not in `get_value()`, also the things you didn't like about the `formulation` template do not apply here.
* I don't understand this comment https://github.com/exadg/exadg/pull/587#discussion_r1383122750

Reply:
The comment you did not understand raises the question whether a base class introducing the `get_value/gradient()` interface should be implemented in deal.II. It is just an idea that does not have to be resolved in the present PR.

> I think the description is correct: The interface to access the value is similar to FEEval.

I agree. However, it is not only similar, it needs to be identical to `FEEval`. At the same time, I think `reinit()` and `set_time()` are independent of `FEEval`. (Generally speaking, this is a disadvantage of using templates vs. base classes and dynamic polymorphism. The definition of interfaces is implicit and somewhat hard to extract from the code, since there is no base class to jump into and see what the required interface is. It can be extracted from the code, but this needs time.) Hence, I consider it important to document/explain this. So I want to make the documentation more precise than it is currently.

Overall, you do not have to change the code as suggested. I want to discuss aspects with pros and cons, and maybe also get other developers involved in the discussion. In the end, I am willing to merge a version that you consider final from your side. When then still have the discussions as a reference for upcoming PRs (by you or by other developers).

Reply:
> A question would then indeed be whether it makes sense that dealii provides a base/interface class `FEFaceEvaluationInterface` with `get_value()/get_gradient()` only. 

This is to some extent related to other discussions we had previously (regarding the "inefficiency" of the `FEEvaluation` interface and the associated `for` loops), where in fact being able to separate the operations at quadrature points from the data containers and the loop index would be helpful. If we consider adding another interface, what you desire as an interface would be possible.

One caveat is that the `get_value()/get_gradient()` cannot be virtual functions because this is one of those places where inlining does have a tangible effect and preventing inlining via function calls would be clearly visible. The other effect preventing a simple change into a `FEEvaluationInterface` as a base class is the complicated inheritance logic present in `FEEvaluation` already now, where the templates on the number of components are used to define different types for the `get_value()`, `get_gradient()` etc. interfaces, as a means to make as intuitive user interfaces as possible. Changing the order by which the classes are set up without introducing virtual functions is probably taking a considerable effort. I think that concentrating the effort on a **new class for the quadrature point operation**, trying to address the various limitations we learned in the past and focusing on usability on various architectures for the future, would be the best option forward. It could then naturally pick up the interface necessary here. We would still need to find a more suitable mechanism than dynamic polymorphism, but I think we are approaching the right problem.

Reply:
Regarding the "complicated inheritance logic present", the class to look at is `FEEvaluationAccess` and its various specializations, https://www.dealii.org/developer/doxygen/deal.II/classFEEvaluationAccess.html

Reply:
I agree that we should evolve the ExaDG code for now and accept that this will be somewhat a temporary solution. I think we will still get ideas for the design of dealii/matrixfree from such developments in ExaDG .